RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.cc

```diff
@@ -0,0 +1,229 @@
+                                               (float)pset.getParameter<double>("quadrupletMinPt"),
+                                               (float)pset.getParameter<double>("quadrupletMaxZip")}};
+  }
+
+}  // namespace
+
+using namespace std;
+
+CAHitNtupletGeneratorOnGPU::CAHitNtupletGeneratorOnGPU(const edm::ParameterSet& cfg, edm::ConsumesCollector& iC)
+    : m_params(cfg.getParameter<bool>("onGPU"),
+               cfg.getParameter<unsigned int>("minHitsPerNtuplet"),
+               cfg.getParameter<unsigned int>("maxNumberOfDoublets"),
+               cfg.getParameter<bool>("useRiemannFit"),
+               cfg.getParameter<bool>("fit5as4"),
+               cfg.getParameter<bool>("includeJumpingForwardDoublets"),
+               cfg.getParameter<bool>("earlyFishbone"),
+               cfg.getParameter<bool>("lateFishbone"),
+               cfg.getParameter<bool>("idealConditions"),
+               cfg.getParameter<bool>("fillStatistics"),
+               cfg.getParameter<bool>("doClusterCut"),
+               cfg.getParameter<bool>("doZ0Cut"),
+               cfg.getParameter<bool>("doPtCut"),
+               cfg.getParameter<double>("ptmin"),
+               cfg.getParameter<double>("CAThetaCutBarrel"),
+               cfg.getParameter<double>("CAThetaCutForward"),
+               cfg.getParameter<double>("hardCurvCut"),
+               cfg.getParameter<double>("dcaCutInnerTriplet"),
+               cfg.getParameter<double>("dcaCutOuterTriplet"),
+               makeQualityCuts(cfg.getParameterSet("trackQualityCuts"))) {
+#ifdef DUMP_GPU_TK_TUPLES
+  printf("TK: %s %s % %s %s %s %s %s %s %s %s %s %s %s %s %s\n",
+         "tid",
+         "qual",
+         "nh",
+         "charge",
+         "pt",
+         "eta",
+         "phi",
+         "tip",
+         "zip",
+         "chi2",
+         "h1",
+         "h2",
+         "h3",
+         "h4",
+         "h5");
+#endif
+
+  if (m_params.onGPU_) {
+    // allocate pinned host memory only if CUDA is available
+    edm::Service<CUDAService> cs;
+    if (cs and cs->enabled()) {
+      cudaCheck(cudaMalloc(&m_counters, sizeof(Counters)));
+      cudaCheck(cudaMemset(m_counters, 0, sizeof(Counters)));
+    }
+  } else {
+    m_counters = new Counters();
+    memset(m_counters, 0, sizeof(Counters));
+  }
+}
+
+CAHitNtupletGeneratorOnGPU::~CAHitNtupletGeneratorOnGPU() {
+  if (m_params.onGPU_) {
+    // print the gpu statistics and free pinned host memory only if CUDA is available
+    edm::Service<CUDAService> cs;
+    if (cs and cs->enabled()) {
+      if (m_params.doStats_) {
+        // crash on multi-gpu processes
+        CAHitNtupletGeneratorKernelsGPU::printCounters(m_counters);
+      }
+      cudaFree(m_counters);
+    }
+  } else {
+    if (m_params.doStats_) {
+      CAHitNtupletGeneratorKernelsCPU::printCounters(m_counters);
+    }
+    delete m_counters;
+  }
+}
+
+void CAHitNtupletGeneratorOnGPU::fillDescriptions(edm::ParameterSetDescription& desc) {
```

Comment:
Actuallty, `CAHitNtupletGeneratorOnGPU` is a bare class and not a plugin module. https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorOnGPU.h#L24
It is merely used by the (plugin) class `CAHitNtupletCUDA` as a member in https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletCUDA.cc#L42. As a consequence, the methods `CAHitNtupletGeneratorOnGPU::fillDescriptions()` and `CAHitNtupletGeneratorOnGPU::fillDescriptionsLabel()` are not directly related to a framework defined interface (besides the homonymy). We could give them different names to avoid the confusion.

Reply:
Strange: to me I t seems that will increase the confusion.

Reply:
The `fillPSetDescription()` is used also elsewhere for similar helper classes without any plugin mechanism. Motivation for the naming is that the framework-called `fillDescriptions(edm:::ConfigurationDescriptions&)` can be used to fill several alternative configurations for a module, whereas a function taking `edm::ParameterSetDescription` can fill only one PSet.